RFC/src/RFC-0173_Versioning.md

```diff
@@ -0,0 +1,123 @@
+WHETHER IN CONTRACT, STRICT LIABILITY OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+## Language
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", 
+"NOT RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in 
+[BCP 14](https://tools.ietf.org/html/bcp14) (covering RFC2119 and RFC8174) when, and only when, they appear in all capitals, as 
+shown here.
+
+## Disclaimer
+
+This document and its content are intended for information purposes only and may be subject to change or update
+without notice.
+
+This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
+community. The release of this document is intended solely for review and discussion by the community of the
+technological merits of the potential system outlined herein.
+
+## Goals
+
+The aim of this Request for Comment (RFC) is to describe the various types of versioning that nodes on the Tari network 
+will use during interaction with other nodes.
+
+## Related Requests for Comment
+
+- [RFC-0710: Tari Communication Network and Network Communication Protocol](RFC-0170_NetworkCommunicationProtocol.md)
+- [RFC-0171: MessageSerialization](RFC-0171_MessageSerialisation.md)
+
+## Description
+
+In a decentralized system the set of nodes on the network will run a variety of software versions as time goes on. Some 
+of these versions will be compatible and others not. For example, if a crucial consensus change is added during a hard 
+fork event. Furthermore, there will be multiple networks running Tari code, i.e. Mainnet vs Testnet. Versioning refers 
+to the strategies we will use for nodes to determine if they can communicate.
+
+Tari will contain three different versioning schemes:
+1. A WireMode byte that precede every low level network packet,
+2. P2P message versions that will accompany every P2P message,
+3. Consensus rules versions that will be exchanged on connection and are included in each block header.
+
+### WireMode byte
+In the Bitcoin P2P protocol messages are preceded by 4 
+[magic values](https://en.bitcoin.it/wiki/Protocol_documentation#Common_structures) or bytes. These values are used to 
+delimit when a new message starts in a byte stream and also are used to indicate which of the Bitcoin networks the node 
+is speaking on, such as TestNet or Mainnet.
+
+Tari message packets are encapsulated using the Noise protocol so we do not need the delimiting functionality of these 
+bytes but Tari will include a single WireMode byte at the beginning of every connection session. This byte will indicate 
+which network a node is communicating on, so that if the counterparty is on a different network it can reject this 
+connection cheaply without having to perform any further operations, like completing the Noise protocol handshake.
```

Comment:
Ideally, we should define the new WireMode mapping.

Currently these are hardcoded values in comms denoting a "mode", all of which are "valid" (in the case of liveness, the byte is always recognised even if it is disabled). 

Essentially we need to move the mapping to the domain layer and allow it to pass the valid byte to comms.   

I suggest the existing `Network` enum is expanded to contain that mapping. 
e.g.
```rust
#[repr(u8)]
enum Network {
   Mainnet = 0x00,
   Testnet1 = 0x01,
   Testnet2 = 0x02,
}

```

Reply:
Sounds good, which Testnet's should be included? Stibbons and Weatherwax?

Reply:
@mikethetike ^

Reply:
I think we should give some space to version testnets and stagenets from mainnets.

Perhaps:
```rust
#[repr(u8)]
enum Network {
   Mainnet = 0x00,
   Stagenet1 = 0x51,
   Testnet1 = 0xa1,
    Testnet2 = 0xa2
}
```

Reply:
Stibbons and Weatherwax are fine as examples